### PR TITLE
Collections

### PR DESCRIPTION
**Pull request by 4P5**
Adds the Collections API which is designed to offload bulk table operations to Java as an optimization.

I've tested all methods for functionality, but some might not handle certain cases. This PR is to allow maintainers to give feedback on the code.

Credit to @TheBunnyMan123 for the initial implementation of the API.

### CODE REVIEW

**Review by TheBunnyMan123 — CHANGES_REQUESTED** [review]
This all looks great, but could you add the new methods to the `wiki` directory in the root of the repository

**Review by PoolloverNathan — APPROVED** [review]
LGTM, need to pad a new release

**Comment by PoolloverNathan** [comment]
misclick sorry

**Comment by PoolloverNathan** [comment]
I'll do the wiki part soon™

**Review by PoolloverNathan — CHANGES_REQUESTED** [review]
actually, not approving — `map` and friends don't have lang entries

File: common/src/main/java/com/thekillerbunny/goofyplugin/lua/CollectionAPI.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        } else if (first.isuserdata(FiguraVec4.class)) {
+            FiguraVec4 quotient = FiguraVec4.of(0, 0, 0, 0);
+            for (int i = 1; i <= tbl.length(); i++) {
+                quotient.divide((FiguraVec4) tbl.get(i).checkuserdata(FiguraVec4.class));
+            }
+            return quotient;
+        } else {
+            throw new LuaError("Unsupported array type");
+        }
+    }
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/lua/CollectionAPI.java:214]: These always return (the appropriate) `0` for vectors.

Comment (PoolloverNathan) [reply common/src/main/java/com/thekillerbunny/goofyplugin/lua/CollectionAPI.java:214]: Here's a patch to default to 1 like doubles do: (idk why it says I replaced the entire code block)
```suggestion
    @LuaWhitelist
    @LuaMethodDoc(
            overloads = {
                    @LuaMethodOverload(
                            argumentTypes = {LuaTable.class},
                            argumentNames = {"tbl"}
                    ),

            },
            value = "collection.sum"
    )
    public Object product(@LuaNotNil LuaTable tbl) {
        if (tbl.length() == 0) {
            return LuaValue.NIL;
        }

        LuaValue first = tbl.get(1);
        if (first.isnumber()) {
            double[] arr = (double[]) CoerceLuaToJava.coerce(tbl, double[].class);
            double product = 1.0;
            for (double v : arr) {
                product *= v;
            }
            return product;
        } else if (first.isuserdata(FiguraVec2.class)) {
            FiguraVec2 product = FiguraVec2.of(1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                product.multiply((FiguraVec2) tbl.get(i).checkuserdata(FiguraVec2.class));
            }
            return product;
        } else if (first.isuserdata(FiguraVec3.class)) {
            FiguraVec3 product = FiguraVec3.of(1, 1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                product.multiply((FiguraVec3) tbl.get(i).checkuserdata(FiguraVec3.class));
            }
            return product;
        } else if (first.isuserdata(FiguraVec4.class)) {
            FiguraVec4 product = FiguraVec4.of(1, 1, 1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                product.multiply((FiguraVec4) tbl.get(i).checkuserdata(FiguraVec4.class));
            }
            return product;
        } else {
            throw new LuaError("Unsupported array type");
        }
    }

    @LuaWhitelist
    @LuaMethodDoc(
            overloads = {
                    @LuaMethodOverload(
                            argumentTypes = {LuaTable.class},
                            argumentNames = {"tbl"}
                    ),

            },
            value = "collection.sum"
    )
    public Object quotient(@LuaNotNil LuaTable tbl) {
        if (tbl.length() == 0) {
            return LuaValue.NIL;
        }

        LuaValue first = tbl.get(1);
        if (first.isnumber()) {
            double[] arr = (double[]) CoerceLuaToJava.coerce(tbl, double[].class);
            double quotient = 1.0;
            for (double v : arr) {
                quotient /= v;
            }
            return quotient;
        } else if (first.isuserdata(FiguraVec2.class)) {
            FiguraVec2 quotient = FiguraVec2.of(1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                quotient.divide((FiguraVec2) tbl.get(i).checkuserdata(FiguraVec2.class));
            }
            return quotient;
        } else if (first.isuserdata(FiguraVec3.class)) {
            FiguraVec3 quotient = FiguraVec3.of(1, 1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                quotient.divide((FiguraVec3) tbl.get(i).checkuserdata(FiguraVec3.class));
            }
            return quotient;
        } else if (first.isuserdata(FiguraVec4.class)) {
            FiguraVec4 quotient = FiguraVec4.of(1, 1, 1, 1);
            for (int i = 1; i <= tbl.length(); i++) {
                quotient.divide((FiguraVec4) tbl.get(i).checkuserdata(FiguraVec4.class));
            }
            return quotient;
        } else {
            throw new LuaError("Unsupported array type");
        }
    }
```

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java [hunk 2]
```diff
@@ -23,7 +24,9 @@ public class FiguraAPIManagerMixin {
 
     static {
         WHITELISTED_CLASSES.add(GoofyAPI.class);
+        WHITELISTED_CLASSES.add(CollectionAPI.class);
```
Comment (PoolloverNathan) [inline common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java:27]: wait, don't we have a thing for this in the plugin code?

File: common/src/main/resources/assets/goofyplugin/lang/en_us.json [hunk 3]
```diff
@@ -29,5 +25,11 @@
...
   "figura.docs.goofy.what_does_bumpscocity_do": "Tells you what bumpscocity does",
   "figura.docs.goofy.get_bumpscocity": "Gets the current bumpscocity",
-  "figura.docs.goofy.set_disable_gui_element": "Sets whether to disable running the\nspecified GUI element"
+  "figura.docs.goofy.set_disable_gui_element": "Sets whether to disable running the\nspecified GUI element",
+  "figura.docs.collection": "Various useful functions for tables",
+  "figura.docs.globals.collection": "The global instance of CollectionAPI",
+  "figura.docs.collection.sum": "Gets the sum of all inputted numbers",
+  "figura.docs.collection.difference": "Gets the difference of all inputted numbers",
+  "figura.docs.collection.product": "Gets the product of all inputted numbers",
+  "figura.docs.collection.quotient": "Gets the quotient of all inputted numbers"
```
Comment (PoolloverNathan) [inline common/src/main/resources/assets/goofyplugin/lang/en_us.json:34]: This is missing `figura.docs.collection.map` and friends.

File: wiki/Collections.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# Collection Operations
+
+For bulk operations on tables, the `collections` API is also provided. This offloads large operations on tables to Java code, allowing you to save instructions and (somewhat) increasing speed.
+
+## Elementwise Operations
+
+* `collection:map(tbl: [a], fn: a → b) → [b]`
+  Runs a function on each element of the array, replacing each element with the return value in-place. Returns the table modified. Similar to `<$>` operator in somelanguages.
```
Comment (PoolloverNathan) [inline wiki/Collections.md:8]: typo
```suggestion
  Runs a function on each element of the array, replacing each element with the return value in-place. Returns the table modified. Similar to `<$>` operator in some languages.
```

File: wiki/Collections.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+* `collection:each(tbl: [a], fn: a → _) → [a]`
+  Like `collection:map`, but does not modify the table. The function's return value is discarded and the original table is returned.
+  ```lua
+  local tbl = {1, 2, 3}
+  local tbl2 = collections:each(tbl, function(a) print(a) end)
+  assert(tbl == tbl2)
+  assert(tbl[1] == 1 and tbl[2] == 2 and tbl[3] == 3)
+  ```
+* `collection:mapEach(tbl: [a], fn: a → [b]) → [b]`
+  Like `collection:map`, but the return values of the functiom are treated as arrays. All such arrays are merged in a newly created array, which is returned. Similar to `>>=` operator in some languages.
```
Comment (PoolloverNathan) [inline wiki/Collections.md:24]:
```suggestion
  Like `collection:map`, but the return values of the function are treated as arrays. All such arrays are merged in a newly created array, which is returned. Similar to `>>=` operator in some languages.
```

File: wiki/Table-Math.md
Comment (PoolloverNathan) [inline wiki/Table-Math.md:1]: should probably delete this file entirely, I'm mentioning this in the collection api page

**Review by PoolloverNathan — DISMISSED** [review]
alright did a real review this time

File: common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java (hunk 2 above)
Comment (TheBunnyMan123) [reply common/src/main/java/com/thekillerbunny/goofyplugin/mixin/FiguraAPIManagerMixin.java:27]: no this is how you do it

File: common/src/main/resources/assets/goofyplugin/lang/en_us.json (hunk 3 above)
Comment (PoolloverNathan) [reply common/src/main/resources/assets/goofyplugin/lang/en_us.json:34]:
```suggestion
  "figura.docs.collection.sum": "Gets the sum of all inputted numbers",
  "figura.docs.collection.difference": "Gets the difference of all inputted numbers",
  "figura.docs.collection.product": "Gets the product of all inputted numbers",
  "figura.docs.collection.quotient": "Gets the quotient of all inputted numbers",
  "figura.docs.collection.map": "Transforms each element of the array in-place",
  "figura.docs.collection.each": "Runs a function on each element of the array",
  "figura.docs.collection.flat_map": "Runs a function on each element and concatentates the resulting arrays",
  "figura.docs.collection.filter": "Returns elements of the array matching a given predicate"
```

**Comment by PoolloverNathan** [comment]
alright we've got the shit now we need to get it in

**Comment by PoolloverNathan** [comment]
tests passed, let's merge

**Review by PoolloverNathan — APPROVED** [review]
alright *now* tests run

**Comment by FokoHetman** [comment]
~~under no situation I should be allowed to review stuff~~
Yeah. seems rather good. My knowledge of figura plugins and everything used here didn't see anything to change really.
I'll test it myself rq and.. probably approved.

**Review by FokoHetman — APPROVED** [review]
tested. It works well and all. Approved

**Pull request merged**